[Hub Generated] Publish private branch 'dev/pratshah/add-cmk-rename-networkacls-add-storage-adx-resourceids'

ramkriz84 opened the pull request:
This is a PR generated at [OpenAPI Hub](https://aka.ms/openapihub). You can view your [work branch via this link](https://openapihub.azure-devex-tools.com/branch/Azure/azure-rest-api-specs/published%2FPratik-Shah%2Fazure-rest-api-specs-pr%2Fdev%2Fpratshah%2Fadd-cmk-rename-networkacls-add-storage-adx-resourceids...main/).
# ARM (Control Plane) API Specification Update Pull Request 

## PR review workflow diagram

Please understand this diagram before proceeding. It explains how to get your PR approved & merged.

![diagram](https://github.com/Azure/azure-rest-api-specs/assets/4429827/b321e1dc-5e07-4940-9cd6-a65ac6729a8b)

[1] [ARM review queue] (for **merge** queues, see [4])  
The PRs are processed by time opened, ascending. Your PR may show up on 2nd or later page. 
If you addressed Step 1 from the diagram and your PR is not showing up in the queue, ensure the label `ARMChangesRequested` 
is removed from your PR. This should cause the label `WaitForARMFeedback` to be added.  
[2] https://aka.ms/azsdk/support/specreview-channel  
[3] [List of SDK breaking changes approvers] in pinned Teams announcement  
[4] [public repo merge queue], [private repo merge queue] (for **ARM review** queue, [1])

If you need further help with anything, see `Getting help` section below.

## Purpose of this PR

What's the purpose of this PR? Check all that apply. This is **mandatory**!

  - [x] New API version. (Such PR should have been generated with [OpenAPI Hub](https://aka.ms/openapihub), per [this wiki doc](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version).)
  - [ ] Update existing version for a new feature. (This is applicable only when you are revising a private preview API version.)
  - [ ] Update existing version to fix swagger quality issues in S360.
  - [ ] Other, please clarify:
    - _edit this with your clarification_

## Due diligence checklist

To merge this PR, you **must** go through the following checklist and confirm you understood 
and followed the instructions by checking all the boxes:

- [x] I confirm this PR is modifying Azure Resource Manager (ARM) related specifications, and not data plane related specifications.
- [x] I have reviewed following [Resource Provider guidelines](https://aka.ms/rpguidelines), including
  [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and
  [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md) (estimated time: 4 hours).  
  I understand this is required before I can proceed to Step 2, "ARM Review", for this PR.

### Breaking changes review (Step 1)

- If the automation determines you have breaking changes, i.e. Step 1 from the diagram applies to you,
  you must follow the [breaking changes process].  
  **IMPORTANT** This applies even if:
  - The tool fails while it shouldn't, e.g. due to runtime exception, or incorrect detection of breaking changes.
  - You believe there is no need for you to request breaking change approval, for any reason. 
    Such claims must be reviewed, and the process is the same.

### ARM API changes review (Step 2)

- If this PR is in purview of ARM review then automation will add the `ARMReview` label.
- If you want to force ARM review, add the label yourself.
- Proceed according to the diagram at the top of this comment.

## Viewing API changes

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

## Suppressing failures

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[Swagger-Suppression-Process](https://aka.ms/azsdk/pr-suppressions) 
to get approval.

## Getting help

- For general PR approval workflow, see the diagram at the top of this comment.
- For guidance on fixing this PR CI check failures, see the hyperlinks provided in given failure 
  and https://aka.ms/ci-fix.
- If the help provided by the previous points is not enough, post to https://aka.ms/azsdk/support/specreview-channel and link to this PR.

[ARM review queue]: https://github.com/search?q=org%3AAzure+is%3Apr+is%3Aopen+label%3AWaitForARMFeedback+-label%3AIDCDevDiv++draft%3Afalse+sort%3Acreated-asc+&type=pullrequests
[List of SDK breaking changes approvers]: https://teams.microsoft.com/l/message/19:0351f5f9404446e4b4fd4eaf2c27448d@thread.skype/1689115217750?tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47&groupId=3e17dcb0-4257-4a30-b843-77f47f1d4121&parentMessageId=1689115217750&teamName=Azure%20SDK&channelName=API%20Spec%20Review&createdTime=1689115217750
[public repo merge queue]: https://github.com/Azure/azure-rest-api-specs/pulls?q=is%3Aopen+is%3Apr+label%3AMergeRequested+draft%3Afalse+sort%3Acreated-asc
[private repo merge queue]: https://github.com/Azure/azure-rest-api-specs-pr/pulls?q=is%3Aopen+is%3Apr+label%3AMergeRequested+-label%3AApproved-OkToMerge+draft%3Afalse+sort%3Acreated-asc
[breaking changes process]: https://eng.ms/docs/cloud-ai-platform/azure-core/azure-core-pm-and-design/trusted-platform-pm-karimb/service-lifecycle-and-actions-team/service-lifecycle-actions-team/apex/media/launchingproductbreakingchanges#breaking-change-process-1

ramkriz84 committed:
add or modify files

TimLovellSmith commented:
> his PR was flagged for introducing a new RP namespace that is not being onboarded with [RPaaS](https://armwiki.azurewebsites.net/rpaas/overview.html). Merging this PR to the main branch will be blocked as RPaaS is required for new RPs. To resolve this and allow the PR to be merged into the main branch, please [use RPaaS to onboard the new RP](https://armwiki.azurewebsites.net/rpaas/gettingstarted.html).

THIS - Plus describe Purpose of this PR. Plus complete due diligence section.

rkmanda commented:
The CI-NewRPNamespaceWithoutRPaaS is a false alarm and may be ignored.

Pratik-Shah commented:
Hello @TimLovellSmith and @rkmanda,

We wanted to get further review on this from you.

_**For this PR, we intend to publish very first version of our service (which is stable a version) in the public repo.**_
 
The related PRs for 2023-11-15 version on the [private arm repo](https://github.com/Azure/azure-rest-api-specs-pr/tree/RPSaaSMaster/specification/networkanalytics/resource-manager/Microsoft.NetworkAnalytics) are as below:

1. Initial PR merge stable version to RPSaaSMaster in private repo. (Merged to private repo) 
   https://github.com/Azure/azure-rest-api-specs-pr/pull/13800

2. Enhancement PR- Added new API to the existing stable ( 2023-11-15)  ( Merged to private repo) 
  https://github.com/Azure/azure-rest-api-specs-pr/pull/13836 

3. (latest) Enhancement PR- Added new feature to the existing (Merged to private repo) 
  https://github.com/Azure/azure-rest-api-specs-pr/pull/14190
	
As there was an architectural change between the preview versions and stable version, it's been decided that the preview version will be deprecated by Jan 2024. Hence, we don't plan preview versions in the public repo.  This makes the stable version the only version which needs to be supported in the public repository.


Questions:
1. Is having only the stable version published to the public repo a valid approach?
2. In this case the stable version itself is a base version for this PR, is this good enough to proceed with the review?
3. The current PR has preview version, should we remove this folder having swagger apis ?

Please let us know your response for the above questions, based on that we can make improvements on this PR.
It would be helpful if we get a slot to talk to you over a call to address this and help the PR proceed further.
Please let us know your convenient time for Monday 18th September 2023.

Thanks,
Pratik 

cc: @ramkriz84 @gaagarwalMS

TimLovellSmith commented:
You do need to do a pass to clean up your randomly generated examples. Remember, these are going to be your public facing documentation. They should be meaningful examples.

Please complete due diligence section.

TimLovellSmith commented:
You also are apparently missing a properly authored 'readme.md' file. Without these automatic validations and SDK generation will not run.

Pratik-Shah commented:
> You do need to do a pass to clean up your randomly generated examples. Remember, these are going to be your public facing documentation. They should be meaningful examples.
> 
> Please complete due diligence section.
Thanks for pointers @TimLovellSmith , will address this.

Just a question here, as oav based examples are randomly generated.
For the public repo is it valid to modify the content of the existing examples files of the public facing examples to make clean ?


> You also are apparently missing a properly authored 'readme.md' file. Without these automatic validations and SDK generation will not run.
Thanks, noted we will add the readme for this version.

Pratik-Shah committed:
Updated examples with meaningful values.
Added Readme for stable version.
removed preview version folder.

Pratik-Shah committed:
Fixed the datatypename naming pattern
Fixed the missing resource Id for datacatalogs pure proxy resource.

TimLovellSmith commented:
> Questions:
> 
> 1. Is having only the stable version published to the public repo a valid approach?
> 2. In this case the stable version itself is a base version for this PR, is this good enough to proceed with the review?
> 3. The current PR has preview version, should we remove this folder having swagger apis ?
> 
> Please let us know your response for the above questions, based on that we can make improvements on this PR. It would be helpful if we get a slot to talk to you over a call to address this and help the PR proceed further. Please let us know your convenient time for Monday 18th September 2023.
> 
> Thanks, Pratik
> 
> cc: @ramkriz84 @gaagarwalMS

Yes, you can have only the stable version published to the public repo, but it will be **impractical** to review for [breaking] changes, compared to the old preview API versions.

In order to enable reviewing the difference between old and new versions, I'd suggest you could publish the stable version to the private repo _as well as_ to the public repo, and asking the reviewer of this PR to check that one for reviewing the difference.

This will be an enabler for breaking changes review, as well as API review.

TimLovellSmith commented:
Also please address existing linter errors: https://github.com/Azure/azure-rest-api-specs/pull/25577/checks?check_run_id=16975910074

Pratik-Shah commented:
Hi @TimLovellSmith, we have addressed the readme and docs comments.

There is a swagger lintDiff check flagged:
[TopLevelResourcesListByResourceGroup](https://github.com/Azure/azure-openapi-validator/blob/main/docs/top-level-resources-list-by-resource-group.md) for DataProductsCatalog.

[TopLevelResourcesListBySubscription](https://github.com/Azure/azure-openapi-validator/blob/main/docs/top-level-resources-list-by-subscription.md) for MetricsIngestionEndpointResources and DataProductsCatalog.

In our usecase DataProductsCatalog and MetricsIngestionEndpointResources which are Singleton resource and requires signoff/exception as these resources are internal to the RP.

The resolution for ModelValidation check is WIP.

We also need some guidance on the below two ci checks on how this can be resolved:
SDK azure-powershell failed: https://github.com/Azure/azure-rest-api-specs/pull/25577/checks?check_run_id=16976000604
SDK azure-resource-manager-schemas failed : https://github.com/Azure/azure-rest-api-specs/pull/25577/checks?check_run_id=16975999444

Pratik-Shah commented:
> > Questions:
> > 
> > 1. Is having only the stable version published to the public repo a valid approach?
> > 2. In this case the stable version itself is a base version for this PR, is this good enough to proceed with the review?
> > 3. The current PR has preview version, should we remove this folder having swagger apis ?
> > 
> > Please let us know your response for the above questions, based on that we can make improvements on this PR. It would be helpful if we get a slot to talk to you over a call to address this and help the PR proceed further. Please let us know your convenient time for Monday 18th September 2023.
> > Thanks, Pratik
> > cc: @ramkriz84 @gaagarwalMS
> 
> Yes, you can have only the stable version published to the public repo, but it will be **impractical** to review for [breaking] changes, compared to the old preview API versions.
> 
> In order to enable reviewing the difference between old and new versions, I'd suggest you could publish the stable version to the private repo _as well as_ to the public repo, and asking the reviewer of this PR to check that one for reviewing the difference.
> 
> This will be an enabler for breaking changes review, as well as API review.

Hello @TimLovellSmith, the stable version is already available private repo which is already signedoff and merge in RPSaaSMaster of private repo.

The related PRs for 2023-11-15 version on the [private arm repo](https://github.com/Azure/azure-rest-api-specs-pr/tree/RPSaaSMaster/specification/networkanalytics/resource-manager/Microsoft.NetworkAnalytics) are as below:

1. Initial PR for 2023-11-15 stable version to RPSaaSMaster in private repo.
    https://github.com/Azure/azure-rest-api-specs-pr/pull/13800
   **- This PR has the base commit of the preview version which was compared with the first iteration of the stable version.**
     -  The initial stable version was merged in to RPSaaSMaster.

2. Enhancement PR- Added new API to the existing stable (2023-11-15) (Merged to private repo)
    https://github.com/Azure/azure-rest-api-specs-pr/pull/13836
    - This PR was to add additional api to stable version.

3. (latest) Enhancement PR- Added new feature to the existing (Merged to private repo)
   https://github.com/Azure/azure-rest-api-specs-pr/pull/14190
   - This PR was to add additional features like to stable version

The history of stable version on private repo can be seen here: https://github.com/Azure/azure-rest-api-specs-pr/pulls?q=is%3Apr+author%3APratik-Shah


As we see the stable version is already merged in RPSaaSMaster in private repo.
Post which the current PR is created using open api publish portal.

Please use the abovementioned PRs to compare stable version with preview versions and enhancements PR.

let me know if this helps you.

Pratik-Shah commented:
We need your help in addressing ci checks:

Hi @isra-fel can you help provide inputs for addressing SDK azure-powershell, How this needs to be resolved. Can you help on this please ?
SDK azure-powershell failed: https://github.com/Azure/azure-rest-api-specs/pull/25577/checks?check_run_id=16976000604


@benbp can you help provide inputs for addressing SDK azure-resource-manager-schemas SDK azure-resource-manager-schemas failed : https://github.com/Azure/azure-rest-api-specs/pull/25577/checks?check_run_id=16975999444

Pratik-Shah commented:
> The CI-NewRPNamespaceWithoutRPaaS is a false alarm. PLease ignore that.

Hi @rkmanda, Since this is a false alarm, can we remove this label ?

susanta-swain-ms commented:
Hi @rkmanda , Regarding the **Swagger ModelValidation failed** errors below, we need you input.
https://github.com/Azure/azure-rest-api-specs/pull/25577/checks?check_run_id=16975911054

INVALID_REQUEST_PARAMETER : api-version 2023-04-30-preview is not equal to swagger version
OBJECT_MISSING_REQUIRED_PROPERTY : Missing required property: api-version
OBJECT_MISSING_REQUIRED_PROPERTY : Missing required property: resourceUri

This Swagger PR is intended for merging of a stable version 2023-11-15 api to public repo but also has a resource type (MetricsIngestionEndpoints) changes with api-version 2023-04-30-preview. Due to this we see the above Model Validation failures because of api-version mismatch between stable and preview version. As MetricsIngestionEndpoints RT is still under preview,  can we remove all reference of it from the stable version swagger ? Or is there any alternate way to segregate them, please let us know your thoughts. Thanks.

Pratik-Shah committed:
Fix for modelvalidation INVALID_REQUEST_PARAMETER
for MetricsIngestionEndpoints

Pratik-Shah committed:
fix for DOUBLE_FORWARD_SLASHES_IN_URL
and  INVALID_FORMAT for MetricsIngestionEndpoints

visingla-ms commented:
Please fix all the errors before adding the label for ARM Review. I can still errors in Linter and Model Validation.

Pratik-Shah committed:
fix for DOUBLE_FORWARD_SLASHES_IN_URL
and  INVALID_FORMAT for MetricsIngestionEndpoints

susanta-swain-ms commented:
> Hi @rkmanda , Regarding the **Swagger ModelValidation failed** errors below, we need you input. https://github.com/Azure/azure-rest-api-specs/pull/25577/checks?check_run_id=16975911054
> 
> INVALID_REQUEST_PARAMETER : api-version 2023-04-30-preview is not equal to swagger version OBJECT_MISSING_REQUIRED_PROPERTY : Missing required property: api-version OBJECT_MISSING_REQUIRED_PROPERTY : Missing required property: resourceUri
> 
> This Swagger PR is intended for merging of a stable version 2023-11-15 api to public repo but also has a resource type (MetricsIngestionEndpoints) changes with api-version 2023-04-30-preview. Due to this we see the above Model Validation failures because of api-version mismatch between stable and preview version. As MetricsIngestionEndpoints RT is still under preview, can we remove all reference of it from the stable version swagger ? Or is there any alternate way to segregate them, please let us know your thoughts. Thanks.

We had a discussion internally and planned to push the MetricsIngestionEndpoints RT changes under the stable API version as well. Later when we change the MetricsIngestionEndpoints resource type it will go under a different api-version.

Pratik-Shah commented:
> Hi @TimLovellSmith, we have addressed the readme and docs comments.
> 
> There is a swagger lintDiff check flagged: [TopLevelResourcesListByResourceGroup](https://github.com/Azure/azure-openapi-validator/blob/main/docs/top-level-resources-list-by-resource-group.md) for DataProductsCatalog.
> 
> [TopLevelResourcesListBySubscription](https://github.com/Azure/azure-openapi-validator/blob/main/docs/top-level-resources-list-by-subscription.md) for MetricsIngestionEndpointResources and DataProductsCatalog.
> 
> In our usecase DataProductsCatalog and MetricsIngestionEndpointResources which are Singleton resource and requires signoff/exception as these resources are internal to the RP.
> 
> The resolution for ModelValidation check is WIP.
> 
> We also need some guidance on the below two ci checks on how this can be resolved: SDK azure-powershell failed: https://github.com/Azure/azure-rest-api-specs/pull/25577/checks?check_run_id=16976000604 SDK azure-resource-manager-schemas failed : https://github.com/Azure/azure-rest-api-specs/pull/25577/checks?check_run_id=16975999444

Hi @TimLovellSmith, we have resolved the model validation check.
Need your help for swagger lintDiff check, as mentioned earlier the DataProductsCatalog and MetricsIngestionEndpointResources are Singleton resource.

Since only one instance of these resource which will ever exists, list by RG and subscription is not going to provide any additional utility in our usecase.

Request to provide an exception/signoff for this check to proceed further.

TimLovellSmith commented:
>Since only one instance of these resource which will ever exists, list by RG and subscription is not going to provide any additional utility in our usecase.

AFAIK you are required to have them anyway, even if they provide no utility, for sake of api consistency.

ramkriz84 commented:
> > Since only one instance of these resource which will ever exists, list by RG and subscription is not going to provide any additional utility in our usecase.
> 
> AFAIK you are required to have them anyway, even if they provide no utility, for sake of api consistency.

Tim, we have cloned what is available in public repo, can this be taken in next iteration?.(first it needs to go in private repo, followed by here). Don't want to block our release because of the same as we have our GA by mid october.

Pratik-Shah commented:
> > > Since only one instance of these resource which will ever exists, list by RG and subscription is not going to provide any additional utility in our usecase.
> > 
> > 
> > AFAIK you are required to have them anyway, even if they provide no utility, for sake of api consistency.
> 
> Tim, we have cloned what is available in private repo, can this be taken in next iteration?.(first it needs to go in private repo, followed by public or can we do this directly in public repo?). Don't want to block our release because of the same as we have our GA by mid October.

Hi @TimLovellSmith , can you provide your input on this ?
We can take up as an enhancement in next iteration and have it as backlog.

Thanks.

Pratik-Shah committed:
Fix to address
1. Swagger LintDiff for DataProductsCatalog Resource
2. Added the examples examples for list by sub and rg
 for DataProductsCatalog
3. minor fix of example title for EMIG

Pratik-Shah committed:
Removed preview versions references and resource
types from swagger.

Pratik-Shah committed:
Fix for typespec validation

Pratik-Shah committed:
fix typespec validation

Pratik-Shah commented:
@TimLovellSmith , we have addressed the Swagger LintDiff and ModelValidation , all the required checks are passing.
Can you please review PR further ?

raosuhas commented on specification/networkanalytics/resource-manager/Microsoft.NetworkAnalytics/stable/2023-11-15/networkanalytics.json:
```diff
@@ -0,0 +1,2954 @@
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.NetworkAnalytics/dataProducts/{dataProductName}/dataTypes/{dataTypeName}/deleteData": {
+      "delete": {
```
>delete [](http://example.com/codeflow?start=7&length=6)

This is incorrect. You should be deleting the root level resource itself , i,e dataTypes/{dataTypeName} according to REST semantics. This is getting flagged in your tests as well 

[https://github.com/Azure/azure-rest-api-specs/pull/25577/checks?check_run_id=17094650089](https://github.com/Azure/azure-rest-api-specs/pull/25577/checks?check_run_id=17094650089)

ramkriz84 replied:
@raosuhas 
This is a specific post action API to our service, and it not for delete of any RT. we have our release mid October so completing the review, ASAP will help us.  if required we can have a short call.
Note: All these changes are reviewed in private repo, we just want to copy the stable API to public repo as part of this PR.

Pratik-Shah replied:
Hi @raosuhas , we have addressed this.
Now the resource DataType has corresponding delete operation and added a post operation for data deletion as well.

Pratik-Shah committed:
Fixed the lintDiff to add
delete operation for datatype proxy resource
Also fixed the deletedata to have post action.
Generated examples for the same.

Pratik-Shah committed:
Fixed referencing of example json
to resolve avacado, model , typespec validations.

Pratik-Shah committed:
fix typespec validation

msyyc commented:
/azp run

dolauli commented:
For Azure PowerShell, I checked the log, and the validation actually passed. The failure is caused by the CI validation framework issue related to github, which is tracked [here](https://github.com/Azure/azure-sdk-tools/issues/6923).

raosuhas commented:
>           "description": "Field to specify storage container sas token."

please mark all secrets as x-ms-secret . please review all your objects and make sure that this is not missed for any secrets

---
Refers to: specification/networkanalytics/resource-manager/Microsoft.NetworkAnalytics/stable/2023-11-15/networkanalytics.json:1924 in fde6829. [](commit_id = fde6829eb48e0f1e5ae22b48ee92a723d0543b49, deletion_comment = False)

raosuhas commented:
>           "202": {

Is this really a long running operation to return the role assignments ?


---
Refers to: specification/networkanalytics/resource-manager/Microsoft.NetworkAnalytics/stable/2023-11-15/networkanalytics.json:1483 in fde6829. [](commit_id = fde6829eb48e0f1e5ae22b48ee92a723d0543b49, deletion_comment = False)

raosuhas commented:
>     "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.NetworkAnalytics/dataProducts/{dataProductName}/addUserRole": {

Can you please explain a bit about this scenario ? What is this user role ? Is this different from the AAD role ?

Why have you decided to model these as post operations for add\remove\list instead of following actual REST semantics ?

---
Refers to: specification/networkanalytics/resource-manager/Microsoft.NetworkAnalytics/stable/2023-11-15/networkanalytics.json:741 in fde6829. [](commit_id = fde6829eb48e0f1e5ae22b48ee92a723d0543b49, deletion_comment = False)

raosuhas commented:
>           "202": {

again is this really a long running operation to generate the token ? Please do a review of the POST actions that you created and see if they really need to be long running ?

---
Refers to: specification/networkanalytics/resource-manager/Microsoft.NetworkAnalytics/stable/2023-11-15/networkanalytics.json:1325 in fde6829. [](commit_id = fde6829eb48e0f1e5ae22b48ee92a723d0543b49, deletion_comment = False)

raosuhas commented:
>               "$ref": "#/definitions/AccountSas"

Can you please rename this to reflect that this is for the properties of the generated sas token input ? AccountSas does not seem to make sense.

---
Refers to: specification/networkanalytics/resource-manager/Microsoft.NetworkAnalytics/stable/2023-11-15/networkanalytics.json:1393 in fde6829. [](commit_id = fde6829eb48e0f1e5ae22b48ee92a723d0543b49, deletion_comment = False)

raosuhas commented:
>               "$ref": "#/definitions/AccountSasToken"

Can you also please explain a bit about this scenario as well ? for which storage account are you generating the sas ? I am mainly trying to understand the RBAC model here. 

---
Refers to: specification/networkanalytics/resource-manager/Microsoft.NetworkAnalytics/stable/2023-11-15/networkanalytics.json:1401 in fde6829. [](commit_id = fde6829eb48e0f1e5ae22b48ee92a723d0543b49, deletion_comment = False)

Pratik-Shah committed:
1. removed EMIG resource , opertions and examples.
2. made below post actions as sync operations
   generateStorageAccountSasToken
  generateStorageContainerSasToken
  rotateKey
  addUserRole
  removeUserRole
  listRolesAssignments

3. Added  x-ms-secret  to the sas token responses
4. refactored examples for the above changes.

Pratik-Shah committed:
Fixed prettier check for example files.

Pratik-Shah commented:
/azp run

Pratik-Shah commented:
> > ```
> >       "description": "Field to specify storage container sas token."
> > ```
> 
> please mark all secrets as x-ms-secret . please review all your objects and make sure that this is not missed for any secrets
> 
> Refers to: specification/networkanalytics/resource-manager/Microsoft.NetworkAnalytics/stable/2023-11-15/networkanalytics.json:1924 in [fde6829](https://github.com/Azure/azure-rest-api-specs/commit/fde6829eb48e0f1e5ae22b48ee92a723d0543b49). [](commit_id = [fde6829](https://github.com/Azure/azure-rest-api-specs/commit/fde6829eb48e0f1e5ae22b48ee92a723d0543b49), deletion_comment = False)

Thanks for pointing, this is addressed. Marked the required secrets as x-ms-secret.

Pratik-Shah commented:
> > ```
> >       "202": {
> > ```
> 
> Is this really a long running operation to return the role assignments ?
> 
> Refers to: specification/networkanalytics/resource-manager/Microsoft.NetworkAnalytics/stable/2023-11-15/networkanalytics.json:1483 in [fde6829](https://github.com/Azure/azure-rest-api-specs/commit/fde6829eb48e0f1e5ae22b48ee92a723d0543b49). [](commit_id = [fde6829](https://github.com/Azure/azure-rest-api-specs/commit/fde6829eb48e0f1e5ae22b48ee92a723d0543b49), deletion_comment = False)

This is addressed. Made list of role assignments as sync operation.

Pratik-Shah commented:
> > ```
> >       "202": {
> > ```
> 
> again is this really a long running operation to generate the token ? Please do a review of the POST actions that you created and see if they really need to be long running ?
> 
> Refers to: specification/networkanalytics/resource-manager/Microsoft.NetworkAnalytics/stable/2023-11-15/networkanalytics.json:1325 in [fde6829](https://github.com/Azure/azure-rest-api-specs/commit/fde6829eb48e0f1e5ae22b48ee92a723d0543b49). [](commit_id = [fde6829](https://github.com/Azure/azure-rest-api-specs/commit/fde6829eb48e0f1e5ae22b48ee92a723d0543b49), deletion_comment = False)

This is addressed. Evaluated and made the below post operations as sync.

- generateStorageAccountSasToken
- generateStorageContainerSasToken
- rotateKey
- addUserRole
- removeUserRole
- listRolesAssignments

Pratik-Shah commented:
> > ```
> >           "$ref": "#/definitions/AccountSas"
> > ```
> 
> Can you please rename this to reflect that this is for the properties of the generated sas token input ? AccountSas does not seem to make sense.
> 
> Refers to: specification/networkanalytics/resource-manager/Microsoft.NetworkAnalytics/stable/2023-11-15/networkanalytics.json:1393 in [fde6829](https://github.com/Azure/azure-rest-api-specs/commit/fde6829eb48e0f1e5ae22b48ee92a723d0543b49). [](commit_id = [fde6829](https://github.com/Azure/azure-rest-api-specs/commit/fde6829eb48e0f1e5ae22b48ee92a723d0543b49), deletion_comment = False)

We plan to go ahead with same name for this release.

Pratik-Shah commented:
> > ```
> > "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.NetworkAnalytics/dataProducts/{dataProductName}/addUserRole": {
> > ```
> 
> Can you please explain a bit about this scenario ? What is this user role ? Is this different from the AAD role ?
> 
> Why have you decided to model these as post operations for add\remove\list instead of following actual REST semantics ?
> 
> Refers to: specification/networkanalytics/resource-manager/Microsoft.NetworkAnalytics/stable/2023-11-15/networkanalytics.json:741 in [fde6829](https://github.com/Azure/azure-rest-api-specs/commit/fde6829eb48e0f1e5ae22b48ee92a723d0543b49). [](commit_id = [fde6829](https://github.com/Azure/azure-rest-api-specs/commit/fde6829eb48e0f1e5ae22b48ee92a723d0543b49), deletion_comment = False)

As briefed over the call, these roles are meant for data product resource.
We will proceed to have post operation for now for this release.

Pratik-Shah commented:
> > ```
> >           "$ref": "#/definitions/AccountSasToken"
> > ```
> 
> Can you also please explain a bit about this scenario as well ? for which storage account are you generating the sas ? I am mainly trying to understand the RBAC model here.
> 
> Refers to: specification/networkanalytics/resource-manager/Microsoft.NetworkAnalytics/stable/2023-11-15/networkanalytics.json:1401 in [fde6829](https://github.com/Azure/azure-rest-api-specs/commit/fde6829eb48e0f1e5ae22b48ee92a723d0543b49). [](commit_id = [fde6829](https://github.com/Azure/azure-rest-api-specs/commit/fde6829eb48e0f1e5ae22b48ee92a723d0543b49), deletion_comment = False)

Closing this, as we have explained the usage of this scenario over the call.
This operation is mainly for the insights storage account for which the token will be generated.

TimLovellSmith commented on specification/networkanalytics/NetworkAnalytics.Management/examples/2023-11-15/DataProducts_ListRolesAssignments_MaximumSet_Gen.json:
```diff
@@ -0,0 +1,31 @@
+{
+  "title": "DataProducts_ListRolesAssignments_MaximumSet_Gen",
+  "operationId": "DataProducts_ListRolesAssignments",
```
>DataProducts_ListRolesAssignments [](http://example.com/codeflow?start=18&length=33)

nit, shouldn't it be 'ListRoleAssignments', not 'ListRolesAssignments'?

Pratik-Shah replied:
This is cloned from private repo. We will first make required change in the private repo and make a fix in public repo in the next release.

TimLovellSmith commented on specification/networkanalytics/NetworkAnalytics.Management/examples/2023-11-15/DataProducts_ListRolesAssignments_MaximumSet_Gen.json:
```diff
@@ -0,0 +1,31 @@
+            "roleAssignmentId": "00000000-0000-0000-0000-00000000000"
+          }
+        ]
+      }
```
now I see an example, I wonder

a) why not follow the usual shape for LIST responses of object holding
{
   "value": [ ... ]
}

b) why include a count? isn't it redundant with the actual list? Unless you have paging or something.

Pratik-Shah replied:
Response for point A. This is not the Top level list response. This is a post call operation where we operate on a managed resource and respond back with list of roles.

Response for point B. It's an auto generated example created using oav tool / typespec  hence maintaining it for consistency.

TimLovellSmith commented on specification/networkanalytics/NetworkAnalytics.Management/examples/2023-11-15/DataProducts_ListRolesAssignments_MaximumSet_Gen.json:
```diff
@@ -0,0 +1,31 @@
+    "subscriptionId": "00000000-0000-0000-0000-00000000000",
+    "resourceGroupName": "aoiresourceGroupName",
+    "dataProductName": "dataproduct01",
+    "body": {}
```
>body [](http://example.com/codeflow?start=5&length=4)

why is there an empty body in the example?

Pratik-Shah replied:
We are query a managed resource hence we don't require any input from user here.

TimLovellSmith commented on specification/networkanalytics/NetworkAnalytics.Management/examples/2023-11-15/DataProducts_Delete_MaximumSet_Gen.json:
```diff
@@ -0,0 +1,18 @@
+    "dataProductName": "dataproduct01"
+  },
+  "responses": {
+    "202": {
```
>202 [](http://example.com/codeflow?start=5&length=3)

delete examples often also usually include an empty 200 response if you follow ARM guidelines for delete responses

```
    "200": {}
```

TimLovellSmith replied:
In this case if you only have long-running delete, I guess I can see an argument for it always completing with 204.

Pratik-Shah replied:
200 response is a vialoation for LRO for delete.
Hence the current response code stays valid.

TimLovellSmith commented on specification/networkanalytics/resource-manager/Microsoft.NetworkAnalytics/stable/2023-11-15/networkanalytics.json:
```diff
@@ -0,0 +1,2571 @@
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.NetworkAnalytics/dataProducts/{dataProductName}/listRolesAssignments": {
+      "post": {
```
>post [](http://example.com/codeflow?start=7&length=4)

I forgot. Why isn't this list operation allowed by GET ting again? Don't you want resource readers to be able to see it?

Pratik-Shah replied:
This is not the Top-level list response. This is a post call operation where we operate on a managed resource and respond back with list of roles.

TimLovellSmith commented on specification/networkanalytics/resource-manager/Microsoft.NetworkAnalytics/stable/2023-11-15/networkanalytics.json:
```diff
@@ -0,0 +1,2571 @@
+      "required": [
+        "startTimeStamp",
+        "expiryTimeStamp",
+        "ipAddress"
```
>ipAddress [](http://example.com/codeflow?start=9&length=9)

should it be required? This isn't usually required for creating SAS uris.

Pratik-Shah replied:
We are creating a Storage account SAS token here, hence we have followed the convention how it's done for storage account.

TimLovellSmith commented on specification/networkanalytics/resource-manager/Microsoft.NetworkAnalytics/stable/2023-11-15/networkanalytics.json:
```diff
@@ -0,0 +1,2571 @@
+        },
+        "ipAddress": {
+          "type": "string",
+          "description": "Ip Address"
```
>Address [](http://example.com/codeflow?start=29&length=7)

Since this is a required property, I think documenting it + its importance more fully could be a good idea!

Pratik-Shah replied:
This is cloned from private repo. We will first make required change in the private repo and make a fix in public repo in the next release.

TimLovellSmith commented on specification/networkanalytics/resource-manager/Microsoft.NetworkAnalytics/stable/2023-11-15/networkanalytics.json:
```diff
@@ -0,0 +1,2571 @@
+        },
+        "fileAccessUrl": {
+          "type": "string",
+          "description": "Url to consume file type.",
```
>Url [](http://example.com/codeflow?start=26&length=3)

URL is an acronym and acronyms should be capitalized in the context of human-oriented docs, descriptions, titles.

Pratik-Shah replied:
This is cloned from private repo. We will first make required change in the private repo and make a fix in public repo in the next release.

TimLovellSmith commented on specification/networkanalytics/resource-manager/Microsoft.NetworkAnalytics/stable/2023-11-15/networkanalytics.json:
```diff
@@ -0,0 +1,2571 @@
+    },
+    "AccountSasToken": {
+      "type": "object",
+      "description": "Details of storage account sas token .",
```
>sas [](http://example.com/codeflow?start=49&length=3)

SAS is an acronym, and acronyms should be capitalized in human-oriented / doc context

Pratik-Shah replied:
This is cloned from private repo. We will first make required change in the private repo and make a fix in public repo in the next release.

TimLovellSmith commented on specification/networkanalytics/resource-manager/Microsoft.NetworkAnalytics/stable/2023-11-15/networkanalytics.json:
```diff
@@ -0,0 +1,2571 @@
+            "description": "The content of the action request",
+            "required": true,
+            "schema": {
+              "$ref": "#/definitions/ContainerSaS"
```
>ContainerSaS [](http://example.com/codeflow?start=37&length=12)

The SaS capitalization is unconventional. It should be SAS or Sas, since its short for Shared Access Signature

Pratik-Shah replied:
This is cloned from private repo. We will first make required change in the private repo and make a fix in public repo in the next release.

TimLovellSmith commented on specification/networkanalytics/resource-manager/Microsoft.NetworkAnalytics/stable/2023-11-15/networkanalytics.json:
```diff
@@ -0,0 +1,2571 @@
+          "DataProducts_GenerateStorageAccountSasToken_MaximumSet_Gen": {
+            "$ref": "./examples/DataProducts_GenerateStorageAccountSasToken_MaximumSet_Gen.json"
+          },
+          "DataProducts_GenerateStorageAccountSasToken_MaximumSet_Gen - generated by [MinimumSet] rule_MinimumSet_Gen": {
```
>generated by [ [](http://example.com/codeflow?start=72&length=14)

you won't want the 'generated by' to show up in your example documents

Pratik-Shah replied:
This is cloned from private repo. We will first make required change in the private repo and make a fix in public repo in the next release.

raosuhas commented:
Signing off on this, but  please review\triage and fix the comments made by @TimLovellSmith before requesting merge

Just FYI though Tim , I am approving this since this PR is time sensitive and the exact same one has already been approved in private repo. These should have been communicated back then, but seemed to have been missed.

Pratik-Shah commented:
/pr RequestMerge
